File: app/src/main/java/tetramax/android/MusicService.java
package tetramax.android;

import android.app.Service;
import android.content.Intent;
import android.content.res.AssetFileDescriptor;
import android.media.MediaPlayer;
import android.os.Binder;
import android.os.IBinder;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;


public class MusicService extends Service {
    static class LocalBinder extends Binder {
        final MusicService service;

        LocalBinder(MusicService service) {
            this.service = service;
        }
    }

    private static final String TAG = MusicService.class.getSimpleName();

    private LocalBinder binder = new LocalBinder(this);

    private final MediaPlayer player = new MediaPlayer();
    private final Random random = new Random();
    public String song = "";

    @Override
    public IBinder onBind(Intent intent) {
        Log.i(TAG, "onBind()");
        return binder;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.i(TAG, "onStartCommand()");
        return START_STICKY;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        Log.i(TAG, "onCreate()");
    }

    @Override
    public void onDestroy() {
        Log.i(TAG, "onDestroy()");
        player.release();
        super.onDestroy();
    }

    public void play() {
        if (player.isPlaying()) {
            return;
        }
        final List<String> files = getFiles();
        song = files.get(random.nextInt(files.size()));

        Log.i(TAG, "Playing song " + song);

        try {
            final AssetFileDescriptor descriptor = getAssets().openFd(song);
            player.setDataSource(
                    descriptor.getFileDescriptor(),
                    descriptor.getStartOffset(),
                    descriptor.getLength()
            );
            descriptor.close();
            player.prepare();
        } catch (IOException e) {
            Log.w(TAG, "Could not open file", e);
            return;
        }

        player.setLooping(true);
        player.start();

        // sends a message to the activity
        final Intent intent = new Intent("mplayer");
        intent.putExtra("song", song);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }

    public void stop() {
        if (player.isPlaying()) {
            player.stop();
        }

        player.reset();
        song = "";

        // send a message to the activity
        final Intent intent = new Intent("mplayer");
        intent.putExtra("song", song);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }

    /**
     * Returns the list of mp3 files in the assets folder
     *
     * @return
     */
    private List<String> getFiles() {
        final List<String> mp3s = new ArrayList<>();
        try {
            final String[] files = getAssets().list("");
            if (files == null) {
                return mp3s;
            }

            for (String fileName : files) {
                if (fileName.toLowerCase().endsWith("mp3")) {
                    mp3s.add(fileName);
                }
            }
        } catch (IOException e) {
            Log.w(TAG, "Error while getting files.", e);
        }

        return mp3s;
    }
}

File: app/src/main/java/tetramax/android/MusicService.kt
package tetramax.android

import android.app.Service
import android.content.Intent
import android.media.MediaPlayer
import android.os.Binder
import android.os.IBinder
import android.support.v4.content.LocalBroadcastManager
import android.util.Log
import java.io.IOException
import java.util.*

class MusicService : Service() {
    val player = MediaPlayer()

    // holds the name of the song currently being played
    public var song = ""

    // an implementation of Binder interface
    internal class LocalBinder(val service: MusicService) : Binder()

    // a reference to LocalBinder
    private val binder = LocalBinder(this)

    override fun onBind(intent: Intent?): IBinder = binder

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int = START_STICKY

    override fun onDestroy() {
        player.release()
        super.onDestroy()
    }

    /**
     * Starts the music player playback
     */
    fun play() {
        if (player.isPlaying) {
            return
        }
        getFiles().random().apply {
            try {
                val descriptor = assets.openFd(this)
                player.setDataSource(
                    descriptor.fileDescriptor,
                    descriptor.startOffset,
                    descriptor.length
                )
                descriptor.close()
                player.prepare()
            } catch (e: IOException) {
                Log.w(TAG, "Could not open file", e)
                return
            }
            player.isLooping = true
            player.start()
            song = this

            // display song info
            Log.i(TAG, "Playing song $this")
            broadcastSongName()
        }
    }

    private fun broadcastSongName() {
        val intent = Intent("mplayer") // mplayer is the name of the broadcast
        intent.putExtra("song", song) // song name is added as the parameter
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent) // the broadcast is sent
    }


    /**
     * Stops the music player playback
     */
    fun stop() {
        if (player.isPlaying) {
            player.stop()
        }
        player.reset()
        this.song = " "
        broadcastSongName()
    }

    /**
     * Returns the list of mp3 files in the assets folder
     *
     * @return
     */
    private fun getFiles(): List<String> =
        assets.list("")?.filter { it.toLowerCase(Locale.getDefault()).endsWith("mp3") }
            ?: emptyList()

    companion object {
        private val TAG = MusicService::class.java.simpleName
    }
}